### PR TITLE
Optimize groupby and distinct according to partitioning

### LINKED ISSUE #500: Support `PushIntoSQL` for `DISTINCT` queries
Currently `DISTINCT` queries are being pushed into a local `DupElim` operator before the shuffle, but the local `DupElim` operator could further be pushed entirely into `DbQueryScan` when that behavior is enabled. This could save a lot of memory for DISTINCT queries on small-cardinality attributes of large relations.

### PR DESCRIPTION
**Pull request by bmyerz**
fixes #500 
- [x] #502 
- [x] add eval correctness to the tests

### CODE REVIEW

**Comment by bmyerz** [comment]
Includes toggling on and off pushing grouping groupbys into Postgres. Use `push_sql_grouping=True`. Default is still off because I don't want to performance-regress any existing Myria workloads

**Comment by bmyerz** [comment]
https://uwdb.slack.com/archives/myria/p1456875590000029

File: raco/backends/myria/myria.py [hunk 1]
```diff
@@ -1481,14 +1470,15 @@ def fire(self, op):
 
 class PushIntoSQL(rules.Rule):
 
-    def __init__(self, dialect=None):
+    def __init__(self, dialect=None, push_grouping=False):
```
Comment (senderista) [inline raco/backends/myria/myria.py]: Is there any long-term reason why aggregates should be treated any differently from projections, selections, or joins WRT pushing into SQL, or is this just short-term risk mitigation?

Comment (senderista) [inline raco/backends/myria/myria.py]: Should we add a new checkbox to the myria-web UI to represent this new flag?

File: raco/backends/myria/myria.py [hunk 2]
```diff
@@ -1713,7 +1703,9 @@ def opt_rules(self, **kwargs):
 
         if kwargs.get('push_sql', False):
             opt_grps_sequence.append([
-                PushIntoSQL(dialect=kwargs.get('dialect'))])
+                PushIntoSQL(dialect=kwargs.get('dialect'),
+                            push_grouping=kwargs.get(
+                                'push_sql_grouping', False))])
```
Comment (senderista) [inline raco/backends/myria/myria.py]: Should we add a new checkbox to the myria-web UI to represent this new flag?

File: raco/backends/sql/catalog.py [hunk 3]
```diff
@@ -110,6 +111,8 @@ def _convert_unary_expr(self, cols, expr, input_scheme):
             return func.max(input)
         if isinstance(expr, expression.MIN):
             return func.min(input)
+        if isinstance(expr, expression.SUM):
```
Comment (senderista) [inline raco/backends/sql/catalog.py]: This should be covered by https://github.com/uwescience/raco/issues/502 and removed from this PR.

File: raco/backends/sql/catalog.py [hunk 4]
```diff
@@ -157,7 +160,7 @@ def _get_unary_sql(self, plan):
             return select(clause, from_obj=input)
 
         elif isinstance(plan, algebra.GroupBy):
-            if len(plan.grouping_list) > 0:
+            if (not self.push_grouping) and len(plan.grouping_list) > 0:
                 raise NotImplementedError(
                     "convert aggregate with grouping to sql -- Myria faster")
```
Comment (senderista) [inline raco/backends/sql/catalog.py]: Can we improve the error message to be more informative?

File: raco/backends/sql/catalog.py [hunk 5]
```diff
@@ -169,6 +172,9 @@ def _get_unary_sql(self, plan):
                 return sel
             return sel.group_by(*g)
 
+        elif isinstance(plan, algebra.Distinct):
```
Comment (senderista) [inline raco/backends/sql/catalog.py]: I want to be sure I understand how this translates to SQL. So something like
`[from R emit distinct foo, bar]`
translates to
```
SELECT DISTINCT S.* FROM (
  SELECT foo, bar FROM R
) AS S;
```
?

File: raco/myrial/optimizer_tests.py [hunk 6]
```diff
@@ -1003,3 +1004,75 @@ def test_projecting_join_maintains_partitioning(self):
         # (in general, info could be h($0) && h($2)
         self.assertEquals(pp.partitioning().hash_partitioned,
                           frozenset([AttIndex(0)]))
+
+    def test_no_shuffle_for_partitioned_distinct(self):
```
Comment (senderista) [inline raco/myrial/optimizer_tests.py]: What about the generic case where shuffles cannot be removed but the local aggregate operator (before the shuffle) can be pushed into SQL?

File: raco/rules.py [hunk 7]
```diff
@@ -832,6 +832,17 @@ def do_transfer(op):
...
 
+    @staticmethod
+    def check_no_shuffle(op):
+        """Check if no shuffle is needed"""
+
+        # Get an array of position references to columns in the child scheme
+        child_scheme = op.input.scheme()
+        group_fields = [expression.toUnnamed(ref, child_scheme)
+                        for ref in op.grouping_list]
+        return len(group_fields) > 0 \
```
Comment (senderista) [inline raco/rules.py]: Nitpick: I prefer to avoid backslash continuations in favor of grouping parentheses.

File: raco/rules.py [hunk 8]
```diff
@@ -808,8 +808,8 @@ class DecomposeGroupBy(Rule):
     The local half of the aggregate before the shuffle step, whereas the remote
     half runs after the shuffle step.
 
-    TODO: omit this optimization if the data is already shuffled, or
-    if the cardinality of the grouping keys is high.
+    TODO: omit this optimization if
```
Comment (senderista) [inline raco/rules.py]: Very happy we're extending the no-shuffle optimization to aggregates!

File: raco/backends/myria/myria.py (hunk 1 above)
Comment (bmyerz) [inline raco/backends/myria/myria.py]: I'm being sensitive to the observation documented in the code that grouping in Myria might be faster.

File: raco/backends/myria/myria.py (hunk 2 above)
Comment (bmyerz) [inline raco/backends/myria/myria.py]: https://github.com/uwescience/myria-web/issues/309

File: raco/backends/sql/catalog.py (hunk 5 above)
Comment (bmyerz) [inline raco/backends/sql/catalog.py]:
```sql
SELECT DISTINCT * 
FROM (SELECT...)
```

Is that what you expect? I have two tests in `optimizer_tests` for this on a single query. Feel free to propose more queries.

File: raco/backends/myria/myria.py (hunk 1 above)
Comment (senderista) [inline raco/backends/myria/myria.py]: Frankly I'm more concerned at this point about memory usage than performance, but being conservative is fine.

File: raco/backends/sql/catalog.py (hunk 5 above)
Comment (senderista) [inline raco/backends/sql/catalog.py]: Yeah that's what I expected, I just wanted to be sure. Obviously Postgres can optimize away the subquery.

File: raco/myrial/optimizer_tests.py (hunk 6 above)
Comment (bmyerz) [inline raco/myrial/optimizer_tests.py]: done

File: raco/rules.py (hunk 7 above)
Comment (bmyerz) [inline raco/rules.py]: done

File: raco/rules.py (hunk 8 above)
Comment (bmyerz) [inline raco/rules.py]: :+1:

File: raco/myrial/optimizer_tests.py (hunk 6 above)
Comment (senderista) [inline raco/myrial/optimizer_tests.py]: That looks great, thanks!

**Comment by senderista** [comment]
LGTM. This is ridiculously OCD, but thought I'd mention that you can eliminate the square brackets in generator expressions like `dict([(k, v) for ...])` (generators don't need to be materialized as list comprehensions).

**Comment by bmyerz** [comment]
ready for merge

**Pull request merged**